### PR TITLE
SVG Icons!

### LINKED ISSUE #365: Add support for SVG icons
And possibly deprecate the icon font.

See https://github.com/blog/2112-delivering-octicons-with-svg

For the same reasons Github states there, the only option for us would be inline SVGs (we need to color them on-the-fly).

This approach would fix:

- Blurriness of some icons at certain browser widths in Chrome & Safari

Caveats:

- Blueprint needs to work across apps / frameworks, so we can't use Github's trick of a Rails helper to inject inline SVGs.
- Given that, we'd have to create React components and Angular directives and `<insert-view-component-in-todays-hot-new-JS-library>` to make usage even remotely concise & reasonable.
- We'd probably also want to provide a separate `@blueprintjs/icons` package with the icon font implementation, in case that works best for certain use cases.

Other notes:

- we can use the SVG CSS property `fill: currentColor` to make icons inherit text colors from a container element, just like the icon font.

### PR DESCRIPTION
**Pull request by giladgray**
#### Checklist

- [x] 🔥 Update documentation
- [x] how to handle the CSS API now? 
  - Buttons, all the way down
  - Input Group CSS API examples
- [x] improve example
- [x] `Callout` still uses icon classes because it relies on `::before` and `position` trickery 
  - ~~can't resolve this without an API break~~  totally did it #2060 
- [x] `MenuItem` icon color

#### Changes proposed in this pull request:

- `generate-icons-source` script now generates `iconSvgPaths.tsx` file with `export const IconSvgPaths{16,20}: Record<IconName, string[]>` where value is array of `<path d="..." />` strings
- 🔥  `IconName` only includes short names. `LegacyIconName` extends it with long names.
  - I made this change for the `Record` type above to support indexing
  - **open to a less-breaking approach**, but i'd also _love_ to drop support for the long names.
- 🔥  core `Icon` component now renders SVG content instead of `span.pt-icon...`
  - removed `iconSize` prop, replaced with `width`/`height`. this isn't totally ideal, I could use some help figuring out the ideal API here.

#### Reviewers should focus on:

- [x] what to do with the CSS icon API?
- [x] how to support 16px / 20px switch?

#### Follow-up work

- 🔥 refactor `iconName?: IconName` prop &rArr; `icon?: IconName | JSX.Element`? #2067 
  - bunch of breaking changes here
  - necessary to support _custom_ SVG icon components
- 🔥 `IconName` or `LegacyIconName` #2066 
  - can we drop support for the long names `pt-icon-time`?
- download links in Icons list should use large icon for 20px download. requires #2066 so I can pass `<Icon iconSize={20} ... />` in that case

### CODE REVIEW

**Comment by giladgray** [comment]
I think I'm going to refactor to `const iconSvgs: Record<IconName, string[]>` where the value is an array of the `<path d="{}">` strings, then i'll render `<svg><paths></svg>` in the component. this'll avoid the `cloneElement` and reduce the size of `iconSvgs.ts` as much as possible (cuz we'll likely need two data files -- 16px and 20px). gotta wait till my [svgo@1.0 types](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23139) publish later today to make this change, though.

File: packages/core/src/components/alert/alert.tsx [hunk 1]
```diff
@@ -69,7 +69,7 @@ export class Alert extends AbstractPureComponent<IAlertProps, {}> {
         return (
             <Dialog className={classNames(Classes.ALERT, className)} isOpen={isOpen} style={style}>
                 <div className={Classes.ALERT_BODY}>
-                    <Icon iconName={iconName} iconSize="inherit" intent={Intent.DANGER} />
+                    <Icon iconName={iconName} width={40} intent={Intent.DANGER} />
```
Comment (llorca) [inline packages/core/src/components/alert/alert.tsx]: You never want to set width and height on an icon, only the size of its bounding box, so what about a `size` prop instead?

File: packages/node-build-scripts/generate-icons-source [hunk 2]
```diff
@@ -4,64 +4,130 @@
...
+ * Loads SVG file for each icon, extracts path strings `d="path-string"`,
+ * and constructs map of icon name to array of path strings.
+ * @param {string} objectName
+ */
+async function buildPathsObject(objectName) {
+    const lines = ICONS_METADATA.map(async icon => {
+        const filepath = path.resolve(__dirname, `../../resources/icons/16px/${icon.className}.svg`);
+        const svg = fs.readFileSync(filepath, "utf-8");
+        const pathStrings = await svgo
+            .optimize(svg, { path: filepath })
```
Comment (llorca) [inline packages/node-build-scripts/generate-icons-source]: 👌 💯

**Comment by llorca** [comment]
What are the downsides of dropping support for the CSS API?

File: packages/core/src/components/icon/icon.tsx [hunk 3]
```diff
@@ -17,50 +17,40 @@ export interface IIconProps extends IIntentProps, IProps {
...
-            getSizeClass(iconSize),
-            Classes.iconClass(iconName),
-            Classes.intentClass(intent),
-            className,
+        const shortName = iconName.replace("pt-icon-", "") as IconName;
+        const classes = classNames(Classes.ICON, Classes.iconClass(iconName), Classes.intentClass(intent), className);
+        return (
+            <svg className={classes} width={width} height={height} viewBox="0 0 16 16">
+                <title>{iconName}</title>
+                {IconSvgPaths[shortName].map((d, i) => <path key={i} d={d} clip-rule="evenodd" fill-rule="evenodd" />)}
```
Comment (llorca) [inline packages/core/src/components/icon/icon.tsx]: I was imagining that the consumer would feed this `IconSvgPaths` as a prop (`iconSet`?), so they can use this component with custom icon sets. Would also solve the 16px / 20px switch, since these two sizes are actually two different sets

File: packages/core/src/components/icon/icon.tsx [hunk 4]
```diff
@@ -17,50 +17,40 @@ export interface IIconProps extends IIntentProps, IProps {
...
-
-        const classes = classNames(
-            getSizeClass(iconSize),
-            Classes.iconClass(iconName),
-            Classes.intentClass(intent),
-            className,
+        const shortName = iconName.replace("pt-icon-", "") as IconName;
+        const classes = classNames(Classes.ICON, Classes.iconClass(iconName), Classes.intentClass(intent), className);
+        return (
+            <svg className={classes} width={width} height={height} viewBox="0 0 16 16">
```
Comment (llorca) [inline packages/core/src/components/icon/icon.tsx]: shouldn't `viewBox` use the size/width props?

File: packages/core/src/components/alert/alert.tsx (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/alert/alert.tsx]: i think `size` is likely the way to go cuz we want them to be squared. 
borrowed this pattern from researching other SVG icons, but it doesn't work super well.

File: packages/core/src/components/icon/icon.tsx (hunk 3 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: oh heck no, we want this to be as easy as possible for _blueprint_ users. i think this `Icon` component should be the Blueprint icon component. if you want to use your own icons, you should do `icon={<MyOwnIcon />}` instead of `icon="time"`.

File: packages/core/src/components/icon/icon.tsx (hunk 4 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: no, viewBox is based on the size of the image itself, so it's either 16 or 20. width/height scale the viewBox.

**Comment by giladgray** [comment]
@llorca main downside of dropping CSS support is that our CSS API sections won't be able to easily show icons (short of copy-pasting the SVG code every time).

File: packages/core/src/components/icon/icon.tsx (hunk 3 above)
Comment (llorca) [reply packages/core/src/components/icon/icon.tsx]: Oh I see, that wasn't clear to me, name feels very generic. Then maybe what you want is something like `<Icon16 />` `<Icon20 />` to reflect the icon set?

Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: What if the default behavior would choose 16 or 20 based on the given `iconSize`? like, the one that requires less scaling? just gotta figure out the math.

`Dialog` is also a generic name 😄

**Comment by giladgray** [comment]
just pushed a commit 5e042e5 that's worth checking out:
1. IconExample on #core/components/icon!
2. this example chooses 16 or 20 based on `iconSize`, so play with that slider!
3. after fiddling with the auto-chooser for a sec, i think it's pretty clearly undesirable behavior. 
most desirable may be the simplest: `iconSize > 16 ? use 20 : use 16`.
4. i'm not sure the cost of two icon sizes (bundle size, maintenance, code complexity) is worth it with SVGs 😐

iconsize-slider

File: packages/node-build-scripts/generate-icons-source [hunk 5]
```diff
@@ -4,64 +4,138 @@
...
-writeLinesToFile("iconStrings.ts", buildTSObject("IconContents", icon => icon.content.replace("\\", "\\u")));
+writeLinesToFile("iconStrings.ts", ...buildTSObject("IconContents", icon => icon.content.replace("\\", "\\u")));
+
+(async () => {
+    // SVG path strings. IIFE to unwrap async.
+    writeLinesToFile(
+        "iconSvgPaths.ts",
+        'import { IconName } from "./iconName";',
+        "",
+        "// tslint:disable:prettier",
```
Comment (adidahiya) [inline packages/node-build-scripts/generate-icons-source]: no need for this; `src/generated` should be excluded from linting

File: packages/core/src/components/icon/icon.tsx [hunk 6]
```diff
@@ -17,50 +17,54 @@ export interface IIconProps extends IIntentProps, IProps {
      * Name of the icon (with or without `"pt-icon-"` prefix).
      * If `undefined`, this component will render nothing.
      */
-    iconName: IconName | undefined;
+    iconName: LegacyIconName | undefined;
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.tsx]: what's the point of distinguishing these from `IconName`?

**Comment by adidahiya** [comment]
feedback on the Icon example:
image
- this should be a `Select` dropdown of icons
- can you put the slider above the resizable icon so it doesn't move as you're using it?

File: packages/core/src/components/icon/icon.tsx [hunk 7]
```diff
@@ -17,50 +17,54 @@ export interface IIconProps extends IIntentProps, IProps {
...
+    private renderSvgPaths(svgPaths: Record<IconName, string[]>) {
+        const paths = svgPaths[this.props.iconName.replace("pt-icon-", "") as IconName];
+        if (paths == null) {
+            return null;
+        }
+        return paths.map((d, i) => <path key={i} d={d} clip-rule="evenodd" fill-rule="evenodd" />);
+    }
+}
+
+function numberDiff(target: number, actual: number) {
```
[the remark this answers is not in the document]
Comment (adidahiya) [reply packages/core/src/components/icon/icon.tsx]: actually, why are we even doing this? visually, we should always use the bigger (20px) SVG for sizes > 20px. there's no benefit to using 16px for those sizes; it looks just as crisp with 20px

File: packages/core/src/components/icon/icon.tsx (hunk 6 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: this is a `[WIP]` hack to make it compile. 
but it also preserves compatibility with the long `pt-icon-time` names.

File: packages/core/src/components/icon/icon.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: yeah this is dead

File: packages/core/src/components/icon/icon.tsx (hunk 6 above)
Comment (adidahiya) [reply packages/core/src/components/icon/icon.tsx]: still doesn't answer the question -- _why_ are we distinguishing short from long names in the type? can't we just accept all of them

Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: I needed a type for the `Record<IconName, string[]>` in `iconSvgPaths` that _didn't_ include the long names so I wouldn't have to duplicate each short name key.* So I [split the old `IconName` enum in two](https://github.com/palantir/blueprint/pull/2028/files#diff-85d2ec5657eb35bcda535552e4063ffeR114) with `LegacyIconName = IconName | ...pt-long-names`

\* - Records with union types are sweet because they enforce that all valid keys have assigned values! Basically ensures that every defined icon has an associated file.

File: package.json [hunk 8]
```diff
@@ -15,7 +15,7 @@
     "copy-landing-app": "cp -rf packages/landing-app/dist/* docs/.",
     "deploy": "gh-pages -d docs -b master",
     "dev:all": "lerna run dev --parallel --scope '!@blueprintjs/{landing-app,table-dev-app}'",
-    "dev:core": "lerna run dev --parallel --scope '@blueprintjs/{core,docs-app}'",
+    "dev:core": "lerna run dev --parallel --scope '@blueprintjs/{core,icons,docs-app}'",
```
Comment (adidahiya) [inline package.json]: this change is unnecessary. we rarely work on icons source. you can run `lerna run dev --parallel --scope '@blueprintjs/icons'` in a separate console.

File: packages/core/src/components/icon/icon.md [hunk 9]
```diff
@@ -41,12 +43,12 @@ The component also accepts all valid HTML props for a `<span>` element.
 ```tsx
 // string literals are supported through IconName union type
 <Icon iconName="cross" />
-<Icon iconName="pt-icon-globe" iconSize="inherit" />
-<Icon iconName="graph" iconSize={20} intent={Intent.PRIMARY} />
+<Icon iconName="pt-icon-globe" width="inherit" />
+<Icon iconName="graph" width={20} intent={Intent.PRIMARY} />
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: this prop change seems wrong.

File: packages/core/src/components/icon/icon.tsx [hunk 10]
```diff
@@ -7,60 +7,63 @@
...
-     * render a 20px icon but inherit `font-size` from its parent.
-     * Constants are exposed for each of these values on the component itself:
-     * `Icon.SIZE_(STANDARD|LARGE|INHERIT)`,
-     * @default 16
+     * Size of the icon, in pixels.
+     * Blueprint contains 16px and 20px SVG icon images,
+     * and chooses the appropriate resolution based on this prop.
      */
-    iconSize?: 16 | 20 | "inherit";
+    iconSize?: number;
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.tsx]: no default?

File: packages/core/src/components/icon/icon.tsx [hunk 11]
```diff
@@ -7,60 +7,63 @@
...
-            return Classes.ICON_LARGE;
-        default:
-            return Classes.ICON;
+    private renderSvgPaths(pathsSize: number) {
+        const svgPaths = pathsSize === Icon.SIZE_STANDARD ? IconSvgPaths16 : IconSvgPaths20;
+        const paths = svgPaths[this.props.iconName.replace("pt-icon-", "") as IconName];
+        if (paths == null) {
+            return null;
+        }
+        return paths.map((d, i) => <path key={i} d={d} fillRule="evenodd" />);
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.tsx]: please document the usage of `fillRule`

File: packages/core/test/icon/iconTests.tsx [hunk 12]
```diff
@@ -14,16 +14,13 @@ import { Classes, Icon, IIconProps, Intent } from "../../src/index";
...
-        ));
+        assertIconSize(<Icon iconName="vertical-distribution" iconSize={Icon.SIZE_STANDARD} />, Icon.SIZE_STANDARD));
 
     it("iconSize=20 renders large size", () =>
-        assertIconClass(<Icon iconName="vertical-distribution" iconSize={Icon.SIZE_LARGE} />, Classes.ICON_LARGE));
+        assertIconSize(<Icon iconName="vertical-distribution" iconSize={Icon.SIZE_LARGE} />, Icon.SIZE_LARGE));
 
-    it("iconSize=inherit renders auto-size", () =>
-        assertIconClass(<Icon iconName="vertical-distribution" iconSize="inherit" />, Classes.ICON));
+    // it("iconSize=inherit renders auto-size", () =>
```
Comment (adidahiya) [inline packages/core/test/icon/iconTests.tsx]: ?

File: packages/core/test/tag-input/tagInputTests.tsx [hunk 13]
```diff
@@ -53,9 +53,10 @@ describe("<TagInput>", () => {
             assert.isTrue(hasClass, errorMessage);
         };
 
-        assertLeftIconHasClass(Classes.ICON_STANDARD, "standard icon");
-        wrapper.setProps({ className: Classes.LARGE });
-        assertLeftIconHasClass(Classes.ICON_LARGE, "large icon");
+        assertLeftIconHasClass(Classes.ICON, "icon");
+        // TODO: large prop
+        // wrapper.setProps({ className: Classes.LARGE });
```
Comment (adidahiya) [inline packages/core/test/tag-input/tagInputTests.tsx]: more commented code

File: package.json (hunk 8 above)
Comment (giladgray) [reply package.json]: in my experience, this change has been very helpful. especially cuz data files come from `icons` package but component comes from `core`. all other dev aliases include package deps, so what's the harm here?

File: packages/core/src/components/icon/icon.tsx (hunk 11 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: @pkwi do you know why this property is here? it came from the original SVG files. is it necessary?

File: packages/core/test/icon/iconTests.tsx (hunk 12 above)
Comment (giladgray) [reply packages/core/test/icon/iconTests.tsx]: the old `Icon` component supported a special `"inherit"` size which worked nicely with fonts. doesn't work so well with SVG because they're special snowflakes and can inherit non-square sizes (which font glyphs couldn't do - always square)

File: packages/core/src/components/icon/icon.tsx (hunk 11 above)
Comment (pkwi) [reply packages/core/src/components/icon/icon.tsx]: yes, this is necessary. `fillRule` is an attribute used in SVG that determines what paths are inside the shape and calculates how the shape is filled. Even-odd counts the amount of paths from the outside, if the path is an odd number, it gets subtracted - it will preserve the holes in overlapping shapes.

File: packages/core/test/icon/iconTests.tsx (hunk 12 above)
Comment (adidahiya) [reply packages/core/test/icon/iconTests.tsx]: sounds like we should remove this test then

File: packages/core/test/popover/popoverTests.tsx [hunk 14]
```diff
@@ -24,9 +24,10 @@ import { Portal } from "../../src/index";
 describe("<Popover>", () => {
     let testsContainerElement: HTMLElement;
     let wrapper: IPopoverWrapper;
-    const onInteractionSpy = sinon.spy();
+    let onInteractionSpy: sinon.SinonSpy;
```
Comment (adidahiya) [inline packages/core/test/popover/popoverTests.tsx]: what's up with the unrelated changes in this file? please revert.

**Comment by adidahiya** [comment]
otherwise, code changes look good overall. CSS API is left intact 👍

File: packages/core/src/components/icon/icon.tsx (hunk 11 above)
Comment (pkwi) [reply packages/core/src/components/icon/icon.tsx]: however, there is a difference between SVG icons exported in Illustrator and Sketch. Illustrator places `fillRule` within the `path` where Sketch is placing it within the `group` container. I dont believe there will be too much harm if we remove it.

File: packages/core/test/popover/popoverTests.tsx (hunk 14 above)
Comment (giladgray) [reply packages/core/test/popover/popoverTests.tsx]: ah semantic merge conflicts. i made this change in two PRs, slightly differently it seems.

File: packages/core/src/components/icon/icon.tsx (hunk 11 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: this is the only place it's defined for the SVG icons. if the `fillRule` property is necessary, then it has to stay right here. 

https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/fill-rule

File: packages/core/src/components/icon/icon.tsx (hunk 6 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.tsx]: @adidahiya what should we do about this interface? are you okay with dropping support for the long names?

**Comment by giladgray** [comment]
@adidahiya @llorca @pkwi updated description with some more outstanding tasks on this feature. please let's discuss.

File: packages/core/src/components/button/_button.scss [hunk 15]
```diff
@@ -87,6 +87,14 @@ Styleguide pt-button
     color: $pt-icon-color;
   }
 
+  // icon-only button
+  .pt-icon:first-child:last-child {
```
Comment (llorca) [inline packages/core/src/components/button/_button.scss]: 🎩

File: packages/core/src/components/button/abstractButton.tsx [hunk 16]
```diff
@@ -127,12 +126,15 @@ export abstract class AbstractButton<T> extends React.Component<React.HTMLProps<
...
         });
 
-        return [
-            loading ? <Spinner className="pt-small pt-button-spinner" key="spinner" /> : undefined,
-            text != null ? <span key="text">{text}</span> : undefined,
-            ...children,
-            <Icon className={Classes.ALIGN_RIGHT} iconName={rightIconName} key="icon" />,
-        ];
+        return (
+            <>
```
Comment (llorca) [inline packages/core/src/components/button/abstractButton.tsx]: so. nice.

**Comment by llorca** [comment]
Monster work, killing it dude 🔥 

Didn't QE everything yet, I'll need some more time. So far I have these comments:
- I'd remove `(none)`, doesn't seem useful for the example. Would be nice to have a default icon selected
screen shot 2018-01-31 at 8 10 25 pm

- in menu items with `intent`, the icon on the left should become white
screen shot 2018-01-31 at 8 13 47 pm

- the icon button on the right side of the input group has both `width` and `height` inaccurate
screen shot 2018-01-31 at 8 13 38 pm

> refactor `iconName?: IconName` prop ⇒ `icon?: IconName | JSX.Element`?

that is 🔑, it is a reason why this PR exists in the first place 👍

File: packages/core/src/components/icon/icon.md [hunk 17]
```diff
@@ -4,15 +4,62 @@
     See [**Icons**](#icons) for a searchable list of all available UI icons.
 </div>
 
+<div class="pt-callout pt-intent-primary pt-icon-info-sign">
+    <h5>SVG icons in 2.0</h5>
+    Blueprint 2.0 introduced SVG icon support throughout the ecosystem, and moved icon resources to a separate
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: nit: no need for the comma here (I noticed your prose style does this a lot and I generally don't mind for code comments, but I'm more picky about it in actual documentation).

Comment (adidahiya) [reply packages/core/src/components/icon/icon.md]: also, "throughout the ecosystem" is a bit of a stretch. I would remove that phrase entirely.

File: packages/core/src/components/icon/icon.md [hunk 18]
```diff
@@ -4,15 +4,62 @@
...
+    __@blueprintjs/icons__ package. The `Icon` component now renders SVG paths and the icon classes are no longer
+    used by any Blueprint React component. Icon font support has been preserved but should be considered legacy,
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: "... but should be considered a legacy feature that will be removed in a future major version."

File: packages/core/src/components/icon/icon.md [hunk 19]
```diff
@@ -4,15 +4,62 @@
...
+The `Icon` component is available in the __@blueprintjs/core__ package.
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+Use the `<Icon>` component to easily render __SVG icons__ in React. The `iconName` prop is typed
+such that editors can offer autocomplete for known icon names. The optional `iconSize` prop ensures
+you'll never forget a sizing class and clarifies the expected width and height of the icon element.
+The component also accepts all valid HTML props for an `<svg>` element.
+
+Data files in the __@blueprintjs/icons__ package provide SVG path information for Blueprint's 300+ icons
+for 16px and 20px grids. The `iconName` prop dictates which SVG is rendered, and `iconSize` determines
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: another unnecessary comma here

File: packages/core/src/components/icon/icon.md [hunk 20]
```diff
@@ -4,15 +4,62 @@
...
+which pixel grid is used: `iconSize >= 20` will use the 20px grid and smaller icons will use the 16px grid.
+
+```tsx
+// string literals are supported through IconName union type
+<Icon iconName="cross" />
+<Icon iconName="pt-icon-globe" iconSize={20} />
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: probably shouldn't encourage the long names with `pt-` if you want to eventually remove them

File: packages/core/src/components/icon/icon.md [hunk 21]
```diff
@@ -4,15 +4,62 @@
...
+<Icon iconName="add" onClick={this.handleAdd} onKeyDown={this.handleAddKeys}>
+```
+
+@interface IIconProps
+
 @## CSS API
 
+<div class="pt-callout pt-intent-warning pt-icon-warning-sign">
+    <h5>Icon fonts are legacy in 2.0</h5>
+    Blueprint's icon fonts should be considered legacy, and we plan to remove them in a future major version.
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.md]: use the same phrasing I suggested above

File: packages/core/src/components/icon/icon.tsx [hunk 22]
```diff
@@ -7,60 +7,73 @@
...
+    public static readonly SIZE_STANDARD = 16;
+    public static readonly SIZE_LARGE = 20;
 
     public render() {
         if (this.props.iconName == null) {
             return null;
         }
-        const { className, iconName, intent, iconSize = Icon.SIZE_STANDARD, ...restProps } = this.props;
+        const { className, iconName: iconNameProp, iconSize = Icon.SIZE_STANDARD, intent, ...svgProps } = this.props;
+        const iconName = iconNameProp.replace("pt-icon-", "") as IconName;
```
Comment (adidahiya) [inline packages/core/src/components/icon/icon.tsx]: nit: leave the prop as `iconName` (also do this destructuring on the first line of the function, so you can use it on L46) and make this variable named `normalizedIconName`

**Comment by pkwi** [comment]
Amazing stuff @giladgray!

Couple of smaller issues:

- Both download links use 16px icons. It should be respectively - 16px icon preview for the 16px SVG download and 20px preview for the 20px download, as we had before.

screen shot 2018-02-01 at 3 42 34 pm

- Adding to @llorca point - for the dark theme - both hovered and selected menu item, the icon on the left should become white - as in [this example](http://blueprintjs.com/docs/v1/#core/components/menu.section-headers)

screen shot 2018-02-01 at 4 05 39 pm

**Comment by adidahiya** [comment]
> refactor iconName?: IconName prop ⇒ icon?: IconName | JSX.Element?

separate PR please

> can we drop support for the long names pt-icon-time?

can you elaborate? we'd still support it in the CSS API, right? does this mean all usages of `iconName={IconClasses...}` would break? what else is likely to break? I'm looking for a comprehensive list.

File: packages/core/src/components/icon/icon.md (hunk 17 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.md]: sometimes, i use commas to indicate where to pause, for maximal dramatic effect. 🎭

**Comment by giladgray** [comment]
@adidahiya yes, dropping long names precisely means that `iconName={IconClasses.*}` would break. I'm down to add `IconNames.*`, though the `IconName` union type effectively provides this in your editor. thoughts?

👍  separate PR for `icon` prop

**Comment by adidahiya** [comment]
ok, in that case I think we can make the break. `iconName` should always be unprefixed.

An `IconNames` object would be nice for consistency with other string enums in our API, but isn't strictly necessary for 2.0.

File: packages/core/src/components/icon/icon.md (hunk 19 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.md]: i actually thought long and hard about this one

File: packages/core/src/components/icon/icon.md (hunk 20 above)
Comment (giladgray) [reply packages/core/src/components/icon/icon.md]: follow-up PR. still technically supported.

**Comment by giladgray** [comment]
i checked off completed issues in comments above.
@pkwi - 20px download link can't be implemented currently. will follow up in #2066.
@llorca - `IconSelect` is generally lackluster at the moment. let's improve it in follow-ups.

**Comment by adidahiya** [comment]
things that still use the icon font:

- .pt-tag-remove
- navbar??
  image
- table column reorder handle
  image



other bugs:

- toasts
  image
- this dialog icon looks too dark?
  image
- Select package is somehow in Core components (probably unrelated to this PR, but it's not present on the live docs
  image
- numeric input regression
  image
